### PR TITLE
Minable payload drop rate manipulation

### PR DESCRIPTION
**Pull request by Amazinite**
**Feature:** This PR implements the prospecting feature from #6050.

## Feature Details
This PR adds a new `prospecting` weapon attribute that increases a counter on any minable asteroid being hit. This value increases the drop rate of a minable asteroid according to the equation `(1. - dropRate) / (1 + toughness / prospecting)`, where `dropRate` is the default drop rate of the payload, `prospecting` is the amount of "prospecting damage" that was applied to the asteroid before it was destroyed, and `toughness` is an attribute of the minable's payload that determines how resistant it is to having its drop rate increased.

This PR also adds new customization to minable payloads.
```
minable "asteroid"
	payload "Iron"
		"max drops" 40 # Default 1
		"drop rate" 0.25 # Default  0.25
		"toughness" 500 # Default 1
```
The maximum payload size, or max drops, can still be defined as a number after the payload outfit (e.g. `payload "Iron" 40`), but this also adds a drop rate attribute (defaults to 0.25) and the toughness attribute described above. This means that a minable with multiple payloads could have each payload affected differently by the same prospecting weapon.

As an example, say a mining laser has a hull damage per frame of 1.9 and a prospecting per frame of 0.1, while a silicon asteroid has a hull of 350. It takes 350/1.9 = 185 frames for a single mining laser to destroy an asteroid. In that time, it will have applied 185 * 0.1 = 18.5 prospecting to the asteroid. If the payload toughness is 37, this would mean that the drop rate of a silicon asteroid destroyed by a mining laser would go from 25% to .25 + .75 / (1 + 37 / 18.5) = 50%. Given that the prospecting applied to an asteroid is a factor of how quickly it was destroyed versus how much prospecting was applied per frame, using multiple mining lasers doesn't further increase the drop rate of an asteroid, only increasing the speed at which the asteroid is destroyed.

Note that a weapon may provide a negative prospecting value to push the drop rate of a minable payload below its default. This may be used to create weapons which mine asteroids very rapidly, but at the cost of reduced yields.

## UI Screenshots
N/A

## Usage Examples
Prospecting is a weapon attribute.
```
outfit "Mining Laser"
	category "Guns"
	cost 108000
	thumbnail "outfit/unknown"
	"mass" 20
	"outfit space" -20
	"weapon capacity" -20
	"gun ports" -1
	weapon
		sprite "projectile/mining laser"
			"frame rate" 1
		sound "laser"
		"hit effect" "beam laser impact"
		"inaccuracy" .6
		"velocity" 200
		"lifetime" 1
		"reload" 1
		"firing energy" 1.5
		"firing heat" 2.3
		"shield damage" 0.4
		"hull damage" 2.9
		"prospecting" .153
```

## Testing Done
Added 0.5 prospecting to a Slicer and modified iron asteroids to have a payload size of 100 and toughness of 192. Mined five asteroids with the Slicer and five without. Recorded the amount of iron harvested from each asteroid.
* Slicer (w/ prospecting): 45, 35, 39, 48, 36
* Other (w/o prospecting): 21, 23, 27, 23, 27

### Automated Tests Added
N/A

## Performance Impact
N/A

### CODE REVIEW

**Comment by Amazinite** [comment]
Need to recalculate new toughness values for the minable asteroids.

File: source/Minable.cpp [hunk 1]
```diff
@@ -43,25 +43,27 @@ void Minable::Load(const DataNode &node)
...
-		else if(child.Token(0) == "hull" && child.Size() >= 2)
+		else if(key == "hull" && hasValue)
 			hull = child.Value(1);
-		else if(child.Token(0) == "random hull" && child.Size() >= 2)
+		else if(key == "random hull" && hasValue)
 			randomHull = max(0., child.Value(1));
-		else if((child.Token(0) == "payload" || child.Token(0) == "explode") && child.Size() >= 2)
+		else if(key == "payload" && hasValue)
+			payload.emplace_back(child);
+		else if(key == "explode" && hasValue)
```
Comment (Hurleveur) [inline source/Minable.cpp:63]: ik this isnt the goal of the PR but whilst we're here shouldnt we check hasvalue once at the start of the if, given all of them require it?

File: source/Minable.cpp [hunk 2]
```diff
@@ -248,3 +255,26 @@ const int64_t &Minable::GetValue() const
 {
 	return value;
 }
+
+
+
+Minable::Payload::Payload(const DataNode& node)
```
Comment (Hurleveur) [inline source/Minable.cpp:261]: isnt it supposed to be at the same place in the cpp and h? or is it here bc its a subclass?

**Review by Hurleveur — COMMENTED** [review]
nice and simple PR, I like the improvments you did to the loading and adding a class

File: source/Minable.cpp (hunk 2 above)
Comment (Amazinite) [reply source/Minable.cpp:261]: It was originally private, but I made it public and never moved the implementation.

File: source/Weapon.cpp [hunk 3]
```diff
@@ -302,6 +302,8 @@ void Weapon::LoadWeapon(const DataNode &node)
 				damage[HIT_FORCE] = value;
 			else if(key == "piercing")
 				piercing = max(0., value);
+			else if(key == "prospecting")
+				prospecting = max(0., value);
```
Comment (Amazinite) [inline source/Weapon.cpp:306]:
```suggestion
				prospecting = value;
```

File: source/Minable.cpp [hunk 4]
```diff
@@ -185,16 +214,20 @@ bool Minable::Move(vector<Visual> &visuals, list<shared_ptr<Flotsam>> &flotsam)
...
 		}
-		for(const auto &it : payload)
+		for(const Payload &it : payload)
 		{
-			if(it.second < 1)
+			// Each payload has a default 25% chance of surviving. This
+			// creates a distribution with occasional very good payoffs.
+			double dropRate = it.dropRate;
+			if(!dropRate)
 				continue;
```
Comment (Amazinite) [inline source/Minable.cpp:223]:
```suggestion
```

File: source/Minable.cpp [hunk 5]
```diff
@@ -185,16 +214,20 @@ bool Minable::Move(vector<Visual> &visuals, list<shared_ptr<Flotsam>> &flotsam)
...
-
-			// Each payload object has a 25% chance of surviving. This creates
-			// a distribution with occasional very good payoffs.
-			for(int amount = Random::Binomial(it.second, .25); amount > 0; amount -= Flotsam::TONS_PER_BOX)
+			// Special weapons are capable of increasing this drop rate through
+			// prospecting.
+			if(prospecting > 0. && dropRate < 1.)
+				dropRate += (1. - dropRate) / (1. + it.toughness / prospecting);
```
Comment (Amazinite) [inline source/Minable.cpp:227]:
```suggestion
				dropRate += (1. - dropRate) / (1. + it.toughness / prospecting);
			if(dropRate <= 0.)
				continue;
```

File: source/Minable.h [hunk 6]
```diff
@@ -38,6 +38,21 @@ class Visual;
...
 // ellipse around the system center.
 class Minable : public Body {
+public:
+	class Payload {
+	public:
+		Payload(const DataNode &node);
+
+		const Outfit *outfit;
+		// The maximum number of outfits that this payload can drop.
+		int dropSize = 1;
```
Comment (Hurleveur) [inline source/Minable.h:48]: The variable name could be clearer I think
maxDrop mb?
dropLimit?
maxDropAmount sounds too long for sure

File: source/Minable.h [hunk 7]
```diff
@@ -38,6 +38,21 @@ class Visual;
...
+		// The average percentage of the maximum number that drop.
```
Comment (Hurleveur) [inline source/Minable.h:49]:
```suggestion
		// The average percentage of the maximum number that drops.
```

**Review by Hurleveur — COMMENTED** [review]
Looks very good I only found a small thing or two

File: source/Minable.h (hunk 7 above)
Comment (Amazinite) [reply source/Minable.h:49]: Sounds fine as is to me.

File: source/Minable.cpp [hunk 8]
```diff
@@ -34,6 +34,31 @@ using namespace std;
...
+
+	for(const DataNode &child : node)
+	{
+		const string &key = child.Token(0);
+		bool hasValue = child.Size() >= 2;
+
+		if(!hasValue)
+			child.PrintTrace("Error: Expected key to have a value:");
+		else if(key == "drop size")
+			dropSize = max<int>(1, child.Value(1));
```
Comment (Amazinite) [inline source/Minable.cpp:50]:
```suggestion
		else if(key == "max drops")
			maxDrops = max<int>(1, child.Value(1));
```

File: source/Minable.cpp [hunk 9]
```diff
@@ -79,7 +108,7 @@ void Minable::Load(const DataNode &node)
 void Minable::FinishLoading()
 {
 	for(const auto &it : payload)
-		value += it.first->Cost() * it.second * 0.25;
+		value += it.outfit->Cost() * it.dropSize * it.dropRate;
```
Comment (Amazinite) [inline source/Minable.cpp:111]:
```suggestion
		value += it.outfit->Cost() * it.maxDrops * it.dropRate;
```

File: source/Minable.cpp [hunk 10]
```diff
@@ -185,16 +214,20 @@ bool Minable::Move(vector<Visual> &visuals, list<shared_ptr<Flotsam>> &flotsam)
...
+			// prospecting.
+			if(prospecting > 0. && dropRate < 1.)
+				dropRate += (1. - dropRate) / (1. + it.toughness / prospecting);
+			if(dropRate <= 0.)
 				continue;
-
-			// Each payload object has a 25% chance of surviving. This creates
-			// a distribution with occasional very good payoffs.
-			for(int amount = Random::Binomial(it.second, .25); amount > 0; amount -= Flotsam::TONS_PER_BOX)
+			for(int amount = Random::Binomial(it.dropSize, dropRate); amount > 0; amount -= Flotsam::TONS_PER_BOX)
```
Comment (Amazinite) [inline source/Minable.cpp:228]:
```suggestion
			for(int amount = Random::Binomial(it.maxDrops, dropRate); amount > 0; amount -= Flotsam::TONS_PER_BOX)
```

File: source/Minable.h (hunk 6 above)
Comment (Amazinite) [inline source/Minable.h:48]:
```suggestion
		int maxDrops = 1;
```

File: source/Minable.cpp [hunk 11]
```diff
@@ -34,6 +34,31 @@ using namespace std;
 
 
 
+Minable::Payload::Payload(const DataNode &node)
+{
+	outfit = GameData::Outfits().Get(node.Token(1));
+	dropSize = (node.Size() == 2 ? 1 : max<int>(1, node.Value(2)));
```
Comment (Amazinite) [inline source/Minable.cpp:40]:
```suggestion
	maxDrops = (node.Size() == 2 ? 1 : max<int>(1, node.Value(2)));
```

File: source/Minable.h (hunk 6 above)
Comment (Amazinite) [reply source/Minable.h]: Changed to maxDrops.

File: source/Minable.h (hunk 7 above)
Comment (Hurleveur) [reply source/Minable.h:49]: Euh sure

**Review by Hurleveur — APPROVED** [review]
I didnt test this but it looks great code wise

**Review by RisingLeaf — APPROVED** [review]

**Pull request merged**